### PR TITLE
feat(exla): convert :optional to xla Call

### PR DESCRIPTION
**Pull request by polvalente**
This is part of our effort to implement compilation boundaries to allow for exchanging of memory and CPU time

### CODE REVIEW

**Comment by seanmor5** [comment]
Nice! Do we have any benchmarks that can compare between using `xla_call` vs. not?

File: exla/lib/exla/defn.ex [hunk 1]
```diff
@@ -1449,6 +1473,31 @@ defmodule EXLA.Defn do
     {EXLA.Builder.build(res), update_outfeed(cache, comp_cache)}
   end
 
+  defp optional_computation(name, arg, expr, state, cache) do
```
Comment (josevalim) [inline exla/lib/exla/defn.ex:1476]: This seems to be the same as `while_computation`. Perhaps we can call it "maybe_token_computation" and reuse it on both?

File: exla/lib/exla/defn.ex [hunk 2]
```diff
@@ -541,15 +541,39 @@ defmodule EXLA.Defn do
...
+        List.to_tuple(computation_args),
+        default_body,
+        state,
+        cache
+      )
+
+    builder =
+      case call_args do
+        [%{builder: builder} | _] ->
+          builder
```
Comment (josevalim) [inline exla/lib/exla/defn.ex:568]: Why do we need this? It should be always state.builder, no?

File: exla/lib/exla/defn.ex [hunk 3]
```diff
@@ -541,15 +541,39 @@ defmodule EXLA.Defn do
...
+
+        _ ->
+          state.builder
+      end
+
+    call = EXLA.Op.call(builder, token, call_args, call_body)
```
Comment (josevalim) [inline exla/lib/exla/defn.ex:574]: We should pass the token as a regular argument (iirc the first argument) in `call_args`, no?

Comment (josevalim) [reply exla/lib/exla/defn.ex:574]: I.e. EXLA.Op.call(builder, computation, args)

**Comment by polvalente** [comment]
> Nice! Do we have any benchmarks that can compare between using `xla_call` vs. not?

@seanmor5 not yet! Any suggestions on how to check that?

**Comment by josevalim** [comment]
@seanmor5 my understanding is that this does not affect performance. It is mostly a way to encapsulate logic to avoid multiple repetitions (the cache is coming later today!).

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**